src/comp.F

```diff
@@ -965,57 +964,71 @@ subroutine comp(tn,o2,o2_nm,o1,o1_nm,he,he_nm,w,mbar,xnmbar,
 !     call mp_periodic_f3d(o1_upd(:,lon0:lon1,lat0-1:lat1+1),
 !    |  lev0,lev1,lon0,lon1,lat0-1,lat1+1)
 !
-! Insure non-negative O2,O:
+! Insure non-negative O2,O,He,N2:
       do lat=lat0,lat1
         do i=lon0,lon1
           do k=lev0,lev1
-            n2(k) = 1.-o2_upd(k,i,lat)-o1_upd(k,i,lat)-he_upd(k,i,lat)
+            if (o2_upd(k,i,lat) < small) o2_upd(k,i,lat) = small
+            if (o1_upd(k,i,lat) < small) o1_upd(k,i,lat) = small
+            if (he_upd(k,i,lat) < small) he_upd(k,i,lat) = small
+!
+            if (o2nm_upd(k,i,lat) < small) o2nm_upd(k,i,lat) = small
+            if (o1nm_upd(k,i,lat) < small) o1nm_upd(k,i,lat) = small
+            if (henm_upd(k,i,lat) < small) henm_upd(k,i,lat) = small
+!
+            n2(k,i,lat) =
+     |        1.-o2_upd(k,i,lat)-o1_upd(k,i,lat)-he_upd(k,i,lat)
+            n2nm(k,i,lat) =
+     |        1.-o2nm_upd(k,i,lat)-o1nm_upd(k,i,lat)-henm_upd(k,i,lat)
+!
+            if (n2(k,i,lat) < small) n2(k,i,lat) = small
+            if (n2nm(k,i,lat) < small) n2nm(k,i,lat) = small
           enddo
+        enddo
+      enddo
 !
 ! When O2+O+He~=1 (typically around z=9), N2 will not scale properly with altitudes
 ! This is an ad-hoc correction to let N2 have the right scaling at very high altitudes
+      if (enforce_n2) then
+        do lat=lat0,lat1
+          do i=lon0,lon1
 !
 ! Find the level where N2 is no longer a major species
-          do ktop=lev0,lev1
-            if (n2(ktop) < 0.01) exit ! Typically around z=7
-          enddo
+            do ktop=lev0,lev1
+              if (n2(ktop,i,lat) < 0.01) exit ! Typically around z=7
+            enddo
 ! Use extrapolation above this height to replace the original values
 ! The scaling uses number density instead of mass mixing ratio
-          if (ktop < lev1) then
-            zmtop = zpmid(ktop)
-            n2top = log(n2(ktop)*xnmbar(ktop,i,lat))
-            n2top1 = log(n2(ktop-1)*xnmbar(ktop-1,i,lat))
-            dn2dz = (n2top-n2top1)/dz
-            do k=ktop+1,lev1
-              n2(k) = exp(n2top+dn2dz*(zpmid(k)-zmtop))/xnmbar(k,i,lat)
-            enddo
-          endif
+            if (ktop < lev1) then
+              zmtop = zpmid(ktop)
+              n2top = log(n2(ktop,i,lat)*xnmbar(ktop,i,lat)*rmassinv_n2)
```

Comment:
why times by rmassinv_n2 here?

Reply:
This was to complete the calculation from mass mixing ratio to number density. It only adds a constant multiplier in the equation but doesn't change the result since constant multipliers are not affecting the calculation for logarithmic extrapolation. I put this multiplier back just to make the equation more consistent.